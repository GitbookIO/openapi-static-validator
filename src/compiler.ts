import escodegen from 'escodegen';
import { namedTypes, builders } from 'ast-types';
import { RequestErrorClass, ValidationErrorClass, ValidationErrorIdentifier } from './error';
import { OpenAPIRef, OpenAPISpec } from './types';
import { compileValueSchema } from './compileValueSchema';
import { hash } from './hash';
import { compileValidateRequest } from './compileValidateRequest';

/**
 * Compiler for OpenAPI specs.
 */
export class Compiler {
    private input: OpenAPISpec;

    /** Map of hash to the identifier */
    private hashes: Map<string, string> = new Map();
    /** Map of objects to the identifier */
    private objectHashes: WeakMap<object, string> = new WeakMap();

    /** Counter to get a new identifier */
    private declarationCounter: number = 0;

    /** Map of identifiers defined globally */
    private globalDeclarations: (
        | namedTypes.FunctionDeclaration
        | namedTypes.ClassDeclaration
        | namedTypes.VariableDeclaration
        | namedTypes.ExportNamedDeclaration
    )[] = [
        builders.exportNamedDeclaration(RequestErrorClass),
        builders.exportNamedDeclaration(ValidationErrorClass),
    ];

    /** Map of hashes already processed */
    private processedHashes: Set<string> = new Set();

    constructor(input: OpenAPISpec = {}) {
        this.input = input;
    }

    /**
     * Define a global identifier with a nickname.
     */
    public declareGlobally(
        declaration:
            | namedTypes.FunctionDeclaration
            | namedTypes.ClassDeclaration
            | namedTypes.VariableDeclaration,
    ) {
        this.globalDeclarations.push(declaration);
    }

    /**
     * Declare something globally basded on an input and return an identifier.
     */
    public declareForInput(
        input: any,
        gen: (
            id: namedTypes.Identifier,
        ) =>
            | namedTypes.FunctionDeclaration
            | namedTypes.ClassDeclaration
            | namedTypes.VariableDeclaration,
    ): namedTypes.Identifier {
        const hash = this.hashInput(input);
        const identifier = builders.identifier(hash);

        if (!this.processedHashes.has(hash)) {
            this.processedHashes.add(hash);
            const fn = gen(identifier);
            this.declareGlobally(fn);
        }

        return identifier;
    }

    /**
     * Define a function to validate an input.
     */
    public declareValidationFunction(
        input: object,
        gen: (args: {
            /** Identifier for the value argument being passed to the function */
            value: namedTypes.Identifier;
            /** Identifier for the path argument being passed to the function */
            path: namedTypes.Identifier;
            /** Identifier for the context argument being passed to the function */
            context: namedTypes.Identifier;
            /** Generate an error */
            error: (message: string) => namedTypes.NewExpression;
        }) => namedTypes.BlockStatement['body'],
    ) {
        const pathIdentifier = builders.identifier('path');
        const valueIdentifier = builders.identifier('value');
        const contextIdentifier = builders.identifier('context');

        const error = (message: string) => {
            return builders.newExpression(ValidationErrorIdentifier, [
                pathIdentifier,
                builders.literal(message),
            ]);
        };

        return this.declareForInput(input, (id) => {
            return builders.functionDeclaration(
                id,
                [pathIdentifier, valueIdentifier, contextIdentifier],
                builders.blockStatement(
                    gen({
                        value: valueIdentifier,
                        path: pathIdentifier,
                        context: contextIdentifier,
                        error,
                    }),
                ),
            );
        });
    }

    /**
     * Build the AST from the entire spec.
     */
    public indexAllComponents() {
        // Index all the schema components.
        const schemas = this.input.components?.schemas ?? {};
        Object.values(schemas).forEach((schema) => {
            compileValueSchema(this, schema);
        });
    }

    /**
     * Return the AST for the program.
     */
    public ast() {
        return builders.program([
            ...compileValidateRequest(this, this.input),
            ...this.globalDeclarations,
        ]);
    }

    /**
     * Generate the JS code for the AST.
     */
    public compile() {
        return escodegen.generate(this.ast(), {
            comment: true,
        });
    }

    /**
     * Hash an object and return an identifier name.
     */
    public hashInput(input: any): string {
        const isObject = typeof input === 'object' && input !== null;

        // Fast track for objects
        if (isObject && this.objectHashes.has(input)) {
            return this.objectHashes.get(input)!;
        }

        const hashValue = hash(input);
        if (this.hashes.has(hashValue)) {
            return this.hashes.get(hashValue)!;
        }

        const name = `obj${this.declarationCounter++}`;
        this.hashes.set(hashValue, name);
        if (isObject) {
            this.objectHashes.set(input, name);
        }

        return name;
    }

    /**
     * Resolve a reference to part of the spec.
     * We only support "#/" type of references.
     */
    public resolveRef(ref: OpenAPIRef) {
        const parts = ref.$ref.split('/').slice(1);

        let value: any = this.input;
        for (const part of parts) {
            value = value[part];
            if (value === undefined) {
                throw new Error(`Could not resolve reference ${ref.$ref}`);
            }
        }

        return value;
    }

    /**
     * Resolve a potential ref to a value.
     */
    public resolveMaybeRef<T>(value: T | OpenAPIRef): T {
        if (typeof value === 'object' && value !== null && '$ref' in value) {
            return this.resolveRef(value);
        }

        return value;
    }
}
