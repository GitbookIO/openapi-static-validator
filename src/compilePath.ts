import { namedTypes, builders } from 'ast-types';

import { Compiler } from "./compiler";
import { OpenAPIPath } from "./types";
import { compileValueSchema } from './compileValueSchema';
import { ValidationErrorIdentifier } from './error';
import { compileOperation } from './compileOperation';

/**
 * Compile a path into a function.
 * The value input is an object:
 * {
 *   path: string;
 *   method: string;
 *   body: any;
 *   query: any;
 *   headers: any;
 * }
 */
export function compilePath(compiler: Compiler, pathOperations: OpenAPIPath) {
    return compiler.defineValidationFunction(pathOperations, ({ value, path, error }) => {
        const nodes: namedTypes.BlockStatement['body'] = [];

        Object.entries(pathOperations).forEach(([method, operation]) => {
            const fnOperation = compileOperation(compiler, operation);

            nodes.push(
                builders.ifStatement(
                    builders.binaryExpression(
                        '===',
                        builders.memberExpression(
                            value,
                            builders.identifier('method'),
                        ),
                        builders.literal(method),
                    ),
                    builders.blockStatement([
                        builders.returnStatement(
                            builders.callExpression(fnOperation, [
                                path,
                                value,
                            ]),
                        ),
                    ]),
                )
            )
        });

        nodes.push(
            builders.returnStatement(
                error('method not supported')
            ),
        )

        return nodes;
    });
}
