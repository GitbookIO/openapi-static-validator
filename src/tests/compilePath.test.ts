import { Compiler } from '../compiler';
import { compilePath } from '../compilePath';

test('with get and post', () => {
    const compiler = new Compiler();
    compilePath(compiler, {
        get: {
            parameters: [
                {
                    name: 'foo',
                    in: 'query',
                    schema: {
                        type: 'number',
                    },
                },
            ],
        },
        post: {
            requestBody: {
                required: true,
                content: {
                    'application/json': {
                        schema: {
                            type: 'object',
                            properties: {
                                foo: {
                                    type: 'number',
                                },
                            },
                        },
                    },
                },
            },
        },
    });
    expect(compiler.compile()).toMatchSnapshot();
});
