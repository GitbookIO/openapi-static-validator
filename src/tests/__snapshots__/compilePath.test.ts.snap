// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`with get and post 1`] = `
"class ValidationError {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value) {
    if (value.body !== undefined) {
        return new ValidationError(path, 'body is not allowed');
    }
    return value;
}
function obj4(path, value) {
    if (!(typeof value === 'number')) {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj3(path, value) {
    const keys = new Set(Object.keys(value));
    const value0 = value['foo'];
    if (value0 !== undefined) {
        const result0 = obj4([
            ...path,
            'foo'
        ], value0);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['foo'] = result0;
    }
    if (keys.size > 0) {
        return new ValidationError(path, 'Unexpected properties');
    }
    return value;
}
function obj2(path, value) {
    if (value.body === undefined) {
        return new ValidationError(path, 'body is required');
    }
    const body = obj3([
        ...path,
        'body'
    ], value.body);
    if (body instanceof ValidationError) {
        return body;
    } else {
        value.body = body;
    }
    return value;
}
function obj0(path, value) {
    if (value.method === 'get') {
        return obj1(path, value);
    }
    if (value.method === 'post') {
        return obj2(path, value);
    }
    return new ValidationError(path, 'method not supported');
}"
`;
